src/accountsdb/buffer_pool.zig

```diff
@@ -0,0 +1,1175 @@
+const std = @import("std");
+const sig = @import("../sig.zig");
+const builtin = @import("builtin");
+
+const FileId = sig.accounts_db.accounts_file.FileId;
+
+/// arbitrarily chosen, I believe >95% of accounts will be <= 512 bytes
+const FRAME_SIZE = 512;
+const INVALID_FRAME = std.math.maxInt(FrameIndex);
+const Frame = [FRAME_SIZE]u8;
+/// we can get away with a 32-bit index
+const FrameIndex = u32;
+const FileOffset = u32;
+const FrameOffset = u10; // 0..=FRAME_SIZE
+
+comptime {
+    // assert our FRAME_SIZE fits in FrameOffset
+    const offset: FrameOffset = FRAME_SIZE;
+    _ = offset;
+}
+
+const LinuxIoMode = enum {
+    Blocking,
+    IoUring,
+};
+const linux_io_mode: LinuxIoMode = .IoUring;
+
+const use_io_uring = builtin.os.tag == .linux and linux_io_mode == .IoUring;
+
+const FileIdFileOffset = struct {
+    const INVALID: FileIdFileOffset = .{
+        .file_id = FileId.fromInt(std.math.maxInt(FileId.Int)),
+        // disambiguate from 0xAAAA / will trigger asserts as it's not even.
+        .file_offset = 0xBAAD,
+    };
+
+    file_id: FileId,
+
+    /// offset in the file from which the frame begin
+    /// always a multiple of FRAME_SIZE
+    file_offset: FileOffset,
+    comptime {
+        if (@sizeOf(FileIdFileOffset) != @sizeOf(u64)) unreachable;
+    }
+};
+
+/// Used for obtaining cached reads.
+///
+/// Design details:
+/// Holds a large number of fixed-size "frames".
+/// Frames have an associated reference counter, but this is for tracking open
+/// handles - frames may outlive the associated rc. Frames are considered dead
+/// when are evicted (which may only happen with 0 open handles).
+/// A frame is always alive when it can be found in frame_map; a frame is always
+/// dead when it is in free_list.
+/// A frame dies when its index is evicted from HierarchicalFifo (inside of
+/// evictUnusedFrame).
+pub const BufferPool = struct {
+    pub const FrameMap = std.AutoHashMapUnmanaged(FileIdFileOffset, FrameIndex);
+
+    /// indices of all free frames
+    /// free frames have a refcount of 0 *and* have been evicted
+    free_list: AtomicStack(FrameIndex),
+
+    /// uniquely identifies a frame
+    /// for finding your wanted index
+    // TODO: thread safety?
+    frame_map: FrameMap,
```

Comment:
Should this be under RwMux?

Reply:
fixed 5a30c0d3dfe06ada0478ffabedb19b644cb7d478